lib/temporal/activity/context.rb

```diff
@@ -1,21 +1,57 @@
+require 'temporal/error/failure'
+
 module Temporal
   class Activity
     class Context
       attr_reader :info
 
-      def initialize(info, heartbeat_proc)
+      def initialize(info, heartbeat_proc, shielded: false)
         @thread = Thread.current
         @info = info
         @heartbeat_proc = heartbeat_proc
+        @cancelled = false
+        @shielded = shielded
+        @mutex = Mutex.new
       end
 
       def heartbeat(*details)
         heartbeat_proc.call(*details)
       end
 
+      def shield(&block)
+        # The whole activity is shielded or called from a nested shield
+        return block.call if @shielded
+
+        mutex.synchronize do
```

Comment:
I would suggest using an atomic counter and throwing when it gets back down to zero instead of this approach.

Say two threads both call `ActivityContext.shield` when `@shielded` is false. They may _both_ make it past line 23 above but one is gonna have to wait for another's entire block to run.

Or maybe this shouldn't be thread safe. We are only raising cancelled errors into the thread the activity is running in, not any thread they may start. So why does `shield` have to be thread safe? Arguably this should error if it's not called on the activity thread because the activity thread is the only one affected when not shielding.

Reply:
You're right, `Context#shield` is only expected to be called on the same object from the same thread, so there's no need for thread-safety when it comes to tracking the `@shielded` state.

However it needs to be thread safe against `Context#cancel`, which will indeed be called from a different thread (a reactor thread). This means that you can't get away with just using atomic counters and need a mutex since we're tracking 2 flags — `@shielded` and `@cancelled`. Without the mutex you might end up not raising if the calls overlap.

Reply:
@antstorm  Do we expect the calls to be nested in this method like:

```
  activity.shield do
          ... some processing logic a
       activity.shield do
          ... some processing logic B
        end  
  end
``` 

Because in that case I think we get a deadlock

Reply:
The calls can be nested, yes, and no you will not get a deadlock. There's also a test for this exact case. This is taken care of by an early escape before acquiring the lock

Reply:
If I understand this corerctly the mutex is here to protect the `cancelled` instance variable. If 2 threads come to the critical section at the same time while `@shielded=false || @cancelled=false` then I don't see how this call is protected?

Reply:
> If 2 threads come to the critical section at the same time

2 threads should not come into this call at the same time. I kinda touched on this in my comment.

@antstorm - I wonder if we want an eager check that current thread ID is the same as context thread ID or we fail. Or maybe we should warn. Or maybe we just do nothing. Yeah, I think one of the latter two just in case people have helpers that can be called in and out of thread and have this shielding.

Reply:
@cretz yeah, I think warning them and doing nothing (early return without shielding) makes sense since we have no idea what other threads the user might be running. Here's the change — https://github.com/temporalio/sdk-ruby/pull/90/commits/b5495845584c800e5e2e0b932868f709d9cef15d